sdks/java/extensions/histogram/src/main/java/org/apache/beam/sdk/extensions/histogram/Histogram.java

```diff
@@ -0,0 +1,571 @@
+        BoundsInclusivity boundsInclusivity) {
+      checkArgument(scale > 0.0, "scale should be positive.");
+      checkArgument(growthFactor > 1.0, "growth factor should be greater than 1.0.");
+      checkArgument(
+          numBoundedBuckets > 0, "number of bounded buckets should be greater than zero.");
+      checkArgument(
+          numBoundedBuckets <= Integer.MAX_VALUE - 2,
+          "number of bounded buckets should be less than max value of integer.");
+
+      ImmutableList.Builder<Double> boundsCalculated = new ImmutableList.Builder<>();
+      // The number of bounds is equal to the numBoundedBuckets + 1.
+      for (int i = 0; i <= numBoundedBuckets; i++) {
+        double bound = scale * Math.pow(growthFactor, i);
+        if (Double.isInfinite(bound)) {
+          throw new IllegalArgumentException("the bound has overflown double type.");
+        }
+        boundsCalculated.add(bound);
+      }
+
+      return new AutoValue_Histogram_BucketBounds(boundsCalculated.build(), boundsInclusivity);
+    }
+
+    /**
+     * Like {@link #exponential(double, double, int, BoundsInclusivity)}, but sets
+     * BoundsInclusivity.LOWER_BOUND_INCLUSIVE_UPPER_BOUND_EXCLUSIVE value for the boundsInclusivity
+     * parameter.
+     */
+    public static BucketBounds exponential(
+        double scale, double growthFactor, int numBoundedBuckets) {
+      return exponential(
+          scale,
+          growthFactor,
+          numBoundedBuckets,
+          BoundsInclusivity.LOWER_BOUND_INCLUSIVE_UPPER_BOUND_EXCLUSIVE);
+    }
+
+    /**
+     * Static factory method for defining bounds of linear histogram and calculating bounds based on
+     * the parameters.
+     *
+     * @param offset value of the lower bound for the first bounded bucket.
+     * @param width bucket width.
+     * @param numBoundedBuckets integer determining the total number of bounded buckets within the
+     *     histogram.
+     * @param boundsInclusivity enum value which defines if lower or upper bounds are
+     *     inclusive/exclusive.
+     */
+    public static BucketBounds linear(
+        double offset, double width, int numBoundedBuckets, BoundsInclusivity boundsInclusivity) {
+      checkArgument(width > 0.0, "width of buckets should be positive.");
+      checkArgument(numBoundedBuckets > 0, "number of bounded buckets should be more than zero.");
+      checkArgument(
+          numBoundedBuckets <= Integer.MAX_VALUE - 2,
+          "number of bounded buckets should be less than max value of integer.");
+
+      ImmutableList.Builder<Double> boundsCalculated = new ImmutableList.Builder<>();
+      // The number of bounds is equal to the numBoundedBuckets + 1.
+      for (int i = 0; i <= numBoundedBuckets; i++) {
+        double bound = offset + i * width;
+        if (Double.isInfinite(bound)) {
+          throw new IllegalArgumentException("the bound has overflown double type.");
+        }
+        boundsCalculated.add(bound);
+      }
```

Comment:
I feel like it can certainly be done, but this optimization may take more time and effort to research this topic, which is not directly in the scope of the combiner. It can be a TODO item for future optimizations. Thanks